File: LeetCode/155. 最小栈.js
/**
 * initialize your data structure here.
 */
const data = Symbol("data");
var MinStack = function () {
  this[data] = [];
};

/**
 * @param {number} x
 * @return {void}
 */
MinStack.prototype.push = function (x) {
  if (this[data].length === 0) {
    this[data].push({
      val: x,
      min: x,
    });
    return;
  }
  let topIdx = this[data].length - 1;
  this[data].push({
    val: x,
    min: Math.min(x, this[data][topIdx].min),
  });
};

/**
 * @return {void}
 */
MinStack.prototype.pop = function () {
  this[data].pop();
};

/**
 * @return {number}
 */
MinStack.prototype.top = function () {
  let topIdx = this[data].length - 1;
  return this[data][topIdx].val;
};

/**
 * @return {number}
 */
MinStack.prototype.getMin = function () {
  let topIdx = this[data].length - 1;
  return this[data][topIdx].min;
};

/**
 * Your MinStack object will be instantiated and called as such:
 * var obj = new MinStack()
 * obj.push(x)
 * obj.pop()
 * var param_3 = obj.top()
 * var param_4 = obj.getMin()
 */

File: LeetCode/155. 最小栈.rs
use std::cmp::min;
struct Data{
    val:i32,
    min:i32
}
struct MinStack {
    data:Vec<Data>
}


/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl MinStack {

    /** initialize your data structure here. */
    fn new() -> Self {
        Self{
            data:Vec::new()
        }
    }
    
    fn push(&mut self, x: i32) {
        match self.data.last() {
            None => self.data.push(Data{
                val:x,
                min:x
            }),
            Some(item) => self.data.push(Data{
                val:x,
                min:min(x,self.data[self.data.len()-1].min)
            })
        }
    }
    
    fn pop(&mut self) {
        self.data.pop();
    }
    
    fn top(&self) -> i32 {
        self.data[self.data.len()-1].val
    }
    
    fn get_min(&self) -> i32 {
        self.data[self.data.len()-1].min
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * let obj = MinStack::new();
 * obj.push(x);
 * obj.pop();
 * let ret_3: i32 = obj.top();
 * let ret_4: i32 = obj.get_min();
 */